src/ios/run.ts

```diff
@@ -10,13 +10,90 @@ import {
   IOSRunException,
 } from '../errors';
 import { getOptionValue } from '../utils/cli';
+import { wait } from '../utils/process';
 
+import type { DeviceValues } from './lib';
+import { IOSLibError } from './lib/lib-errors';
 import { getBundleId, unzipIPA } from './utils/app';
 import { getConnectedDevices, runOnDevice } from './utils/device';
+import type { SimulatorResult } from './utils/simulator';
 import { getSimulators, runOnSimulator } from './utils/simulator';
 
 const debug = Debug('native-run:ios:run');
 
+interface IOSRunConfig {
+  udid?: string;
+  devices: DeviceValues[];
+  simulators: SimulatorResult[];
+  appPath: string;
+  bundleId: string;
+  waitForApp: boolean;
+  preferSimulator: boolean;
+}
+
+async function runIpaOrAppFile({
+  udid,
+  devices,
+  simulators,
+  appPath,
+  bundleId,
+  waitForApp,
+  preferSimulator,
+}: IOSRunConfig): Promise<void> {
+  if (udid) {
+    if (devices.find(d => d.UniqueDeviceID === udid)) {
+      await runOnDevice(udid, appPath, bundleId, waitForApp);
+    } else if (simulators.find(s => s.udid === udid)) {
+      await runOnSimulator(udid, appPath, bundleId, waitForApp);
+    } else {
+      throw new IOSRunException(
+        `No device or simulator with UDID "${udid}" found`,
+        ERR_TARGET_NOT_FOUND,
+      );
+    }
+  } else if (devices.length && !preferSimulator) {
+    // no udid, use first connected device
+    await runOnDevice(devices[0].UniqueDeviceID, appPath, bundleId, waitForApp);
+  } else {
+    // use default sim
+    await runOnSimulator(
+      simulators[simulators.length - 1].udid,
+      appPath,
+      bundleId,
+      waitForApp,
+    );
+  }
+}
+
+async function runIpaOrAppFileOnInterval(config: IOSRunConfig): Promise<void> {
```

Comment:
This is a nitpick so @dwieeb I'll defer to you, but I'm not sure the readability gain here is worth it. I'd just kill this outright and move it back into `run()` since `config` is just being passed through and the timeout is hardcoded. If we ever make that configurable from a flag it saves you a refactor.

Reply:
I think we split it out because we're defining inner functions, which is weird to do in a function that already does a lot.

Reply:
You and your functions 😛 Marking resolved.